cudax/include/cuda/experimental/__device/arch_traits.cuh

```diff
@@ -0,0 +1,401 @@
+#endif // no system header
+
+namespace cuda::experimental
+{
+
+namespace detail
+{
+struct arch_common_traits
+{
+  // TODO: Should these be in block_level, grid_level?
+  // Maximum number of threads per block
+  static constexpr int max_threads_per_block = 1024;
+
+  // Maximum x-dimension of a block
+  static constexpr int max_block_dim_x = 1024;
+
+  // Maximum y-dimension of a block
+  static constexpr int max_block_dim_y = 1024;
+
+  // Maximum z-dimension of a block
+  static constexpr int max_block_dim_z = 64;
+
+  // Maximum x-dimension of a grid
+  static constexpr int max_grid_dim_x = cuda::std::numeric_limits<int>::max();
+
+  // Maximum y-dimension of a grid
+  static constexpr int max_grid_dim_y = 64 * 1024 - 1;
+
+  // Maximum z-dimension of a grid
+  static constexpr int max_grid_dim_z = 64 * 1024 - 1;
+
+  // Maximum amount of shared memory available to a thread block in bytes
+  static constexpr int max_shared_memory_per_block = 48 * 1024;
+
+  // Memory available on device for __constant__ variables in a CUDA C kernel in bytes
+  static constexpr int total_constant_memory = 64 * 1024;
+
+  // Warp size in threads
+  static constexpr int warp_size = 32;
+
+  // Maximum number of concurrent grids on the device
+  static constexpr int max_resident_grids = 128;
+
+  // TODO: Not sure if we need these:
+  // true if the device can concurrently copy memory between host and device
+  // while executing a kernel, or false if not
+  static constexpr bool gpu_overlap = true;
+
+  // true if the device can map host memory into CUDA address space
+  static constexpr bool can_map_host_memory = true;
+
+  // true if the device supports executing multiple kernels within the same
+  // context simultaneously, or false if not. It is not guaranteed that multiple
+  // kernels will be resident on the device concurrently so this feature should
+  // not be relied upon for correctness.
+  static constexpr bool concurrent_kernels = true;
+
+  // true if the device supports stream priorities, or false if not
+  static constexpr bool stream_priorities_supported = true;
+
+  // true if device supports caching globals in L1 cache, false if not
+  static constexpr bool global_l1_cache_supported = true;
+
+  // true if device supports caching locals in L1 cache, false if not
+  static constexpr bool local_l1_cache_supported = true;
+
+  // TODO: We might want to have these per-arch
+  // Maximum number of 32-bit registers available to a thread block
+  static constexpr int max_registers_per_block = 64 * 1024;
+
+  // Maximum number of 32-bit registers available to a multiprocessor; this
+  // number is shared by all thread blocks simultaneously resident on a
+  // multiprocessor
+  static constexpr int max_registers_per_multiprocessor = 64 * 1024;
+
+  // Maximum number of 32-bit registers available to a thread
+  static constexpr int max_registers_per_thread = 255;
+};
+} // namespace detail
+
+struct arch_traits_t : public detail::arch_common_traits
```

Comment:
dump question. Is it not weird that `arch_traits_t` can be modified at run-time?

Reply:
Instances are constexpr, so I don't think there would be any memory cost except when its passed around.

The members are not const so I can initialize them in the lambda with proper assignments, I can change it, but the original instances are constexpr, so only a local copy can be modified, which is probably fine 🤷